pkg/omni/repository/repository.go

```diff
@@ -0,0 +1,179 @@
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
+	"github.com/vulcanize/vulcanizedb/pkg/omni/contract"
+	"github.com/vulcanize/vulcanizedb/pkg/omni/types"
+)
+
+// Repository is used to
+type DataStore interface {
+	PersistEvents(info *contract.Contract) error
+}
+
+type dataStore struct {
+	*postgres.DB
+}
+
+func NewDataStore(db *postgres.DB) *dataStore {
+
+	return &dataStore{
+		DB: db,
+	}
+}
+
+// Creates a schema for the contract
+// Creates tables for the watched contract events
+// Persists converted event log data into these custom tables
+func (d *dataStore) PersistEvents(con *contract.Contract) error {
+
+	schemaExists, err := d.CheckForSchema(con.Name)
+	if err != nil {
+		return err
+	}
+
+	if !schemaExists {
+		err = d.CreateContractSchema(con.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	for eventName := range con.Filters {
+		event := con.Events[eventName]
+		if len(event.Logs) == 0 {
+			break
+		}
+
+		tableExists, err := d.CheckForTable(con.Name, eventName)
+		if err != nil {
+			return err
+		}
+
+		if !tableExists {
+			err = d.CreateEventTable(con.Name, event)
+			if err != nil {
+				return err
+			}
+		}
+
+		for id, log := range event.Logs {
+			// Create postgres command to persist any given event
+			pgStr := fmt.Sprintf("INSERT INTO %s.%s ", strings.ToLower(con.Name), strings.ToLower(eventName))
+			pgStr = pgStr + "(vulcanize_log_id, token_name, token_address, event_name, block, tx"
+			var data []interface{}
+			data = append(data,
+				id,
+				strings.ToLower(con.Name),
+				strings.ToLower(con.Address),
+				strings.ToLower(eventName),
+				log.Block,
+				log.Tx)
+
+			counter := 0
+			for inputName, input := range log.Values {
+				// postgres cannot handle custom types, resolve to strings
+				switch input.(type) {
+				case big.Int:
+					var b big.Int
+					b = input.(big.Int)
+					input = b.String()
+				case *big.Int:
+					var b *big.Int
+					b = input.(*big.Int)
+					input = b.String()
+				case common.Address:
+					var a common.Address
+					a = input.(common.Address)
+					input = a.String()         // this also gives us a chance to add any event emitted address
+					con.AddAddress(a.String()) // to a list of token-related addresses, growing it as we go
+				case common.Hash:
+					var h common.Hash
+					h = input.(common.Hash)
+					input = h.String()
+				}
+
+				counter += 1
+				pgStr = pgStr + fmt.Sprintf(", _%s", strings.ToLower(inputName))
+				data = append(data, input)
+			}
+
+			appendStr := "VALUES ($1, $2, $3, $4, $5, $6"
+
+			for i := 0; i < counter; i++ {
+				appendStr = appendStr + fmt.Sprintf(", $%d", i+7)
+			}
+
+			appendStr = appendStr + ") "
+			appendStr = appendStr + "ON CONFLICT (vulcanize_log_id) DO NOTHING"
+			pgStr = pgStr + fmt.Sprintf(") %s", appendStr)
+
+			_, err := d.DB.Exec(pgStr, data...)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
```

Comment:
It's really awesome to have so much dynamic behavior built into this approach to persistence 👍 

I can tell that a lot of thought has already gone into how to do so in a sane and reliable way, but this method is still pretty large - any other ways we could refactor it to cut down on/extract its responsibilities?

Reply:
Thank you! And I agree, this method needs some refactoring.